### PR TITLE
Refactor relocatable shader compilation

### PR DESCRIPTION
**Pull request by s-perron**
To match other work that is being done, we want relocatable shader
compilation to be divided into vertex processing and fragment stages.
The current implementation tries to compile each shader stage separatly.
This is currently a difference without distinction because only handle
CS and VsFs pipelines.

When we will try to handle geometry shader, we want to compile the vertex
shader and the geometry shader together.   We want to refactor the
relocatable shader code to make that easier.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit f3b8148231f71a7a11e31c78082781adbbdc97d8
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 9786fa8c34df6d9baddeff40d331106799fcbb07</li>
<li>SPVGEN: dev 5bf4618ac00d4675f57ca5b571a485bbc4d4fa33</li>
<li>XGL: dev 9b77f6008d4316922cde45baf39d079b4d328893</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 7da5ed485404334756d765594dd151e1ad31af55</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225557)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 30786/56392 (54.6%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 25606/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30785/56392 (54.6%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 25607/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 30917/56381 (54.8%)</li>
<li>Failed: 0/56381 (0.0%)</li>
<li>Not Supported: 25464/56381 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34387/56392 (61.0%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 22005/56392 (39.0%)</li>
</ul></details>


</ul></details>

**Comment by kuhar** [comment]
This rebase test did not change anything, see https://github.com/GPUOpen-Drivers/llpc/pull/1262#issuecomment-854330006

File: include/vkgcDefs.h
Comment (amdrexu) [inline include/vkgcDefs.h:204]: UnlinkedShaderTypeEnum -> UnlinkedShaderStage

File: include/vkgcDefs.h
Comment (amdrexu) [inline include/vkgcDefs.h:206]: ->
UnlinkedStageVertexProcess
UnlinkedStageFragment
UnlinkedStageCompute
UnlinkedStageCount

File: include/vkgcDefs.h
Comment (amdrexu) [inline include/vkgcDefs.h:205]: I don't think this enum is useful unless the start value is non-zero. So currently you can always use 0 as the loop initial value.
Also, if you provide `first`, I think we always need `last`, which matches Vulkan style: UnlinkedStageFirst, UnlinkedStageLast.
And the loop could be:
`for (stage = UnlinkedStageFirst; stage <= UnlinkedStageLast; ++stage)`

**Review by amdrexu — COMMENTED** [review]
LGTM.

**Comment by s-perron** [comment]
V2: Renamed the enum and some variables.  Updated logging and test to match the change from Unlinked shader type to unlinked shader stage.

**Comment by amdvlk-admin** [comment]
### Test summary for commit fd6d5d81b67822584174c06274a33719e3e58275
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 173662704ec193436c8cde216d536775f8cace0a</li>
<li>SPVGEN: dev 79f8e32dd7242d7d34fa0074f761c9d794f11811</li>
<li>XGL: dev 45a98d826d546e6829a021268ff0c599150411cb</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 7da5ed485404334756d765594dd151e1ad31af55</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/225557)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 30786/56392 (54.6%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 25606/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 30785/56392 (54.6%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 25607/56392 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 30917/56381 (54.8%)</li>
<li>Failed: 0/56381 (0.0%)</li>
<li>Not Supported: 25464/56381 (45.2%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34387/56392 (61.0%)</li>
<li>Failed: 0/56392 (0.0%)</li>
<li>Not Supported: 22005/56392 (39.0%)</li>
</ul></details>


</ul></details>

File: llpc/util/llpcUtil.cpp
Comment (kuhar) [inline llpc/util/llpcUtil.cpp:136]: Can you hoist this logic into a lambda or a helper function? The current implementation seems prone to copy-paste-edit errors where someone forgets to update one part of the condition.

File: llpc/util/llpcUtil.h
Comment (kuhar) [inline llpc/util/llpcUtil.h:126]: Maybe `getStageTypeName`? `TypeName` alone sounds ambiguous to me, I would not associate it with shader stages as my first thought

File: llpc/util/llpcUtil.cpp
Comment (kuhar) [inline llpc/util/llpcUtil.cpp:153]: nit: s/`unsigned int`/`unsigned`

File: include/vkgcDefs.h
Comment (kuhar) [inline include/vkgcDefs.h:204]: I think this is treated as `unsigned` elsewhere. If that's the case, it'd be better to be explicit here and have `enum UnlinkedShaderStage : unsigned`.

File: llpc/util/llpcUtil.h
Comment (kuhar) [inline llpc/util/llpcUtil.h:128]: typo in `unlinked`

File: llpc/util/llpcUtil.h
Comment (kuhar) [inline llpc/util/llpcUtil.h:134]: typo in `unlinked`

File: llpc/util/llpcUtil.h
Comment (kuhar) [inline llpc/util/llpcUtil.h:129]: I guess the alternative would be to have something like the enum iterator in LLVM: https://github.com/llvm/llvm-project/blob/1465e7770bcacfb67ca4edc3714cf3a7e8a2a110/llvm/lib/Target/AMDGPU/SIInsertWaitcnts.cpp#L56. This could be added to some utility header in LLPC. This seems like a cleaner abstraction to me compared to loops with casts and these increments, but I don't have a strong preference.

File: llpc/util/llpcUtil.h
Comment (kuhar) [inline llpc/util/llpcUtil.h:123]: nit: s/`unsigned int`/`unsigned`

File: llpc/util/llpcUtil.h
Comment (s-perron) [reply llpc/util/llpcUtil.h:129]: This will be up to AMD, but we should probably get a standard solution in place.  I do not like the casts in the loops either.  To remove the cast for the initialization, the best way is to introduce a `_FIRST` element, but I was asked to remove it (https://github.com/GPUOpen-Drivers/llpc/pull/1264#discussion_r646620996).  I don't  like the template because not all enumerated types can be iterator over that way, and it could be easy to think you are doing something correct when you are not.  I like having those at can be explicitly marked as such.  We could also declare an array containing all of the enums and iterator over the array.

**Comment by s-perron** [comment]
V3: Fixes based on Jakub's comments.  Fixed spelling, changed some variable names, and created a new function to check a shaderInfo array if a particular shader stage exists.

File: llpc/util/llpcUtil.h
Comment (kuhar) [reply llpc/util/llpcUtil.h:129]: We could mark iteratable enums with a tag and only accept that in the utility implementing enum ranges. Something like:
```
// In a utility header:
template <typename E>
struct iteratable_enum;

template <typename E, typename = std::void_t<iteratable_enum<E>>
struct enum_range {

}; ...

// At the enum definition place.
enum MyEnum : unsinged {
X = 0,
...,
Z,
Last = Z
};

template <>
struct iterable_enum<MyEnum> {};
```

Comment (kuhar) [reply llpc/util/llpcUtil.h:129]: But anything more complicated than these increment operators should probably go into a separate PR anyway

**Comment by amdvlk-admin** [comment]
### Test summary for commit 216cbff913c739ebe02f77109ec6416d18eda78e
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 173662704ec193436c8cde216d536775f8cace0a</li>
<li>SPVGEN: dev 79f8e32dd7242d7d34fa0074f761c9d794f11811</li>
<li>XGL: dev 45a98d826d546e6829a021268ff0c599150411cb</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 7da5ed485404334756d765594dd151e1ad31af55</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228104)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31146/57032 (54.6%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25886/57032 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31145/57032 (54.6%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25887/57032 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31278/57008 (54.9%)</li>
<li>Failed: 0/57008 (0.0%)</li>
<li>Not Supported: 25730/57008 (45.1%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34853/57032 (61.1%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 22179/57032 (38.9%)</li>
</ul></details>


</ul></details>

File: llpc/util/llpcUtil.h
Comment (s-perron) [reply llpc/util/llpcUtil.h:129]: Agreed, and I see you open something up for discussion.  I'll mark this conversation as resolved.

File: include/vkgcDefs.h
Comment (trenouf) [inline include/vkgcDefs.h:204]: I guess you might have considered using an enum class here if you had Jakub's enum iterator thing to avoid all the casting in "for" loops.

**Review by trenouf — APPROVED** [review]
LGTM

File: include/vkgcDefs.h
Comment (s-perron) [reply include/vkgcDefs.h:204]: Yes I would have.

**Comment by JaxLinAMD** [comment]
@kuhar @amdrexu, wait for your approval.

**Review by kuhar — APPROVED** [review]

**Review by amdrexu — APPROVED** [review]

**Pull request merged**
